### PR TITLE
cli m005/s01 Port git-service extension from gsd-pi

### PR DESCRIPTION
**Pull request by gannonh**
## What Changed
See slice plan: S01: (no slice plan found)

## Must-Haves
- See slice plan

## Summary by CodeRabbit

* **New Features**
  * Added Git Service for the Kata CLI with branch management, conventional commit message generation, staging, merge operations, and conflict error handling.

* **Tests**
  * Added comprehensive test suite covering branch operations, commit message inference, staging, merge scenarios, and conflict detection.

<h3>Greptile Summary</h3>

This PR ports the `git-service` extension from `gsd-pi` into the Kata CLI, introducing core git primitives — `runGit`, `smartStage`, `commit`, `autoCommitCurrentBranch`, `mergeSliceToMain` — along with supporting types, constants, and a comprehensive integration test suite that exercises these against real temporary git repositories.

The foundational pieces (`runGit`, `inferCommitType`, `buildTaskCommitMessage`, smart staging, branch detection) are well-implemented and well-tested. The top-level merge workflow in `mergeSliceToMain`, however, has two logic bugs and no test coverage:

- **Silent false-success on empty squash** (`git-service.ts` line 483): the return value of `commit()` is discarded. When a squash produces nothing staged (empty or already-merged slice), `commit()` returns `null` but `mergeSliceToMain` still returns a successful `MergeSliceResult` with a `mergedCommitMessage`, misleading callers into thinking a real commit was created.
- **`merge_strategy: "merge"` silently ignored** (`git-service.ts` line 462): `mergeSliceToMain` always executes `git merge --squash` regardless of `prefs.merge_strategy`. The `MergeConflictError` propagates the pref value as `strategy`, creating a contradictory message if the user configured `"merge"`.
- **Negative `maxDescLen`** (`git-service.ts` line 226): if `taskId` is long enough that `68 - type.length - scope.length < 0`, `description.slice(0, maxDescLen - 1)` truncates from the *end* of the string rather than the beginning, producing an over-long subject line.
- **`mergeSliceToMain` and `MergeConflictError` have zero test coverage** in the test file, which is notable given the bugs found above.

<h3>Confidence Score: 3/5</h3>

- Not safe to merge without fixing the silent false-success in mergeSliceToMain and the ignored merge_strategy preference.
- The pure-helper and staging layers are solid and well-tested, but mergeSliceToMain — the most consequential function in the module — has two correctness bugs and no test coverage. The discarded commit() return value means callers can receive a success signal for a merge that produced no commit, which is a silent data-loss risk. Score of 3 reflects solid foundations marred by untested, buggy top-level merge logic.
- apps/cli/src/resources/extensions/kata/git-service.ts (mergeSliceToMain, buildTaskCommitMessage) and apps/cli/src/resources/extensions/kata/tests/git-service.test.ts (missing mergeSliceToMain coverage)

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/cli/src/resources/extensions/kata/git-service.ts | New git service module with solid foundations (runGit, smartStage, inferCommitType) but mergeSliceToMain has two logic bugs: the commit() return value is discarded (silent no-commit on empty squash), and merge_strategy: "merge" preference is silently ignored while MergeConflictError still reports it, creating a misleading mismatch. buildTaskCommitMessage also has an edge case with negative maxDescLen for long taskId scopes. |
| apps/cli/src/resources/extensions/kata/tests/git-service.test.ts | Comprehensive integration tests for pure helpers and staging logic using real temp git repos; however, the most complex exported function (mergeSliceToMain) and the MergeConflictError class have zero test coverage. |

</details>

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Caller
    participant mergeSliceToMain
    participant runGit
    participant commit

    Caller->>mergeSliceToMain: mergeSliceToMain(basePath, milestoneId, sliceId, sliceTitle, prefs)
    mergeSliceToMain->>runGit: getCurrentBranch → sliceBranch
    mergeSliceToMain->>runGit: getMainBranch → mainBranch
    mergeSliceToMain->>runGit: git switch mainBranch
    mergeSliceToMain->>runGit: git merge --squash sliceBranch
    alt Merge conflict
        runGit-->>mergeSliceToMain: throws
        mergeSliceToMain->>runGit: git diff --name-only --diff-filter=U
        runGit-->>mergeSliceToMain: conflictedFiles[]
        mergeSliceToMain-->>Caller: throws MergeConflictError
    else Squash OK
        runGit-->>mergeSliceToMain: success
        mergeSliceToMain->>commit: commit(basePath, squashMsg)
        commit->>runGit: git diff --cached --name-only
        alt Nothing staged (empty slice) ⚠️
            commit-->>mergeSliceToMain: null (return value discarded!)
            mergeSliceToMain-->>Caller: MergeSliceResult (false success)
        else Files staged
            commit->>runGit: git commit -F - (stdin)
            runGit-->>commit: success
            commit-->>mergeSliceToMain: squashMsg
            mergeSliceToMain-->>Caller: MergeSliceResult{branch, mergedCommitMessage, deletedBranch:false}
        end
    end
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: apps/cli/src/resources/extensions/kata/git-service.ts
Line: 483-485

Comment:
**Silent no-commit on empty squash**

`commit()` returns `null` when nothing is staged (e.g. the slice branch has no net diff against `mainBranch`). Here the return value is discarded, so `mergeSliceToMain` reports a successful merge and returns a `mergedCommitMessage` even though **no commit was ever created**. Callers that trust the non-null result to indicate a real commit will have incorrect state.

```suggestion
  const squashMsg = `feat(${milestoneId}/${sliceId}): ${sliceTitle}`;
  const committed = commit(basePath, { message: squashMsg });
  if (!committed) {
    throw new Error(
      `Squash of ${sliceBranch} into ${mainBranch} staged nothing — slice may be empty or already merged`,
    );
  }

  return { branch: sliceBranch, mergedCommitMessage: squashMsg, deletedBranch: false };
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/cli/src/resources/extensions/kata/git-service.ts
Line: 462

Comment:
**`merge_strategy: "merge"` silently ignored**

`GitPreferences.merge_strategy` accepts both `"squash"` and `"merge"`, but `mergeSliceToMain` always calls `git merge --squash` regardless of the preference value. The `MergeConflictError` below even propagates `prefs?.merge_strategy ?? "squash"` in its `strategy` field — so if a user has `merge_strategy: "merge"`, the error would claim the strategy was `"merge"` while the code actually performed a squash. Either the non-squash path should be implemented, or the `merge_strategy` type should be narrowed to only `"squash"` until the regular-merge path is added.

```
// Current — merge_strategy preference is never consulted:
runGit(basePath, ["merge", "--squash", sliceBranch]);

// Should branch on prefs?.merge_strategy:
if ((prefs?.merge_strategy ?? "squash") === "squash") {
  runGit(basePath, ["merge", "--squash", sliceBranch]);
} else {
  runGit(basePath, ["merge", "--no-ff", sliceBranch]);
}
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/cli/src/resources/extensions/kata/git-service.ts
Line: 226

Comment:
**Negative `maxDescLen` for long `taskId` + type combinations**

`maxDescLen` is computed as `68 - type.length - scope.length`. `type` can be up to 8 chars (`"refactor"`), and there is no cap on `taskId`/`scope` length. If `scope.length > 60`, `maxDescLen` goes negative, and `description.slice(0, maxDescLen - 1)` becomes something like `description.slice(0, -10)` — which strips 10 chars from the **end** of the string rather than trimming to the desired length, leaving an over-long subject line.

Adding a guard prevents the subject from exceeding the 72-char budget:

```suggestion
  const maxDescLen = Math.max(1, 68 - type.length - scope.length);
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/cli/src/resources/extensions/kata/tests/git-service.test.ts
Line: 1-20

Comment:
**No tests for `mergeSliceToMain` or `MergeConflictError`**

`mergeSliceToMain` is the most complex exported function in `git-service.ts` — it orchestrates a branch switch, squash-merge, conflict detection, and commit — but the test file imports neither it nor `MergeConflictError`. Given the logic bug identified above (discarded `commit()` return value), and the untested `merge_strategy` branching, this function has zero test coverage. At minimum, the following cases should be covered:

- Happy path: slice branch with commits → squash merged and committed onto `mainBranch`
- Empty/already-merged slice → function should not silently succeed
- Conflict scenario → `MergeConflictError` thrown with correct `conflictedFiles` populated

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: ["feat(S01/T03): imple..."](https://github.com/gannonh/kata/commit/cc7c8846d9d02bb67c197876b58aec054ae1a194)</sub>

> Greptile also left **4 inline comments** on this PR.

### CODE REVIEW

File: apps/cli/src/resources/extensions/kata/git-service.ts [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+      strategy: prefs?.merge_strategy ?? "squash",
+      branch: sliceBranch,
+      mainBranch,
+    });
+  }
+
+  const squashMsg = `feat(${milestoneId}/${sliceId}): ${sliceTitle}`;
+  commit(basePath, { message: squashMsg });
+
+  return { branch: sliceBranch, mergedCommitMessage: squashMsg, deletedBranch: false };
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/kata/git-service.ts]: Fixed. Captured the `commit()` return value and throw when it's `null`:

```ts
const committed = commit(basePath, { message: squashMsg });
if (!committed) {
  throw new Error(
    `Squash of ${sliceBranch} into ${mainBranch} staged nothing — slice may be empty or already merged`,
  );
}
```

An empty-slice squash is an error condition in Kata's workflow (not a silent no-op), so throwing is the correct signal. Added a test case covering this edge case (78 tests passing).

File: apps/cli/src/resources/extensions/kata/git-service.ts [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+  sliceTitle: string,
+  prefs?: GitPreferences,
+): MergeSliceResult {
+  const sliceBranch = getCurrentBranch(basePath);
+  const mainBranch = getMainBranch(basePath, prefs);
+
+  runGit(basePath, ["switch", mainBranch]);
+
+  try {
+    runGit(basePath, ["merge", "--squash", sliceBranch]);
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/kata/git-service.ts:465]: Partially fixed. The core issue is that `MergeConflictError` was reporting `prefs?.merge_strategy ?? "squash"` in its `strategy` field, which could claim `"merge"` was used even though the code always runs `--squash`. Fixed by hardcoding `strategy: "squash"` in the error (accurate — it reflects what actually executed).

```ts
throw new MergeConflictError({
  ...
  strategy: "squash", // always squash — non-squash merge path not yet implemented
  ...
});
```

Also added a comment on the `runGit` call noting that only squash-merge is currently implemented. Full `--no-ff` merge support is a future TODO and out of scope for this port.

File: apps/cli/src/resources/extensions/kata/git-service.ts [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+ * what was actually built), falling back to the task title (what was planned).
+ * The commit type is inferred from the title and one-liner keywords.
+ */
+export function buildTaskCommitMessage(ctx: TaskCommitContext): string {
+  const scope = ctx.taskId;
+  const description = ctx.oneLiner ?? ctx.taskTitle;
+  const type = inferCommitType(ctx.taskTitle, ctx.oneLiner);
+
+  // Truncate description to keep subject line under ~72 chars
+  const maxDescLen = 68 - type.length - scope.length;
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/kata/git-service.ts]: Fixed. Added `Math.max(1, ...)` guard as suggested:

```ts
const maxDescLen = Math.max(1, 68 - type.length - scope.length);
```

This ensures `description.slice(0, maxDescLen - 1)` never receives a negative index (which would strip from the end rather than truncate to the desired length).

File: apps/cli/src/resources/extensions/kata/tests/git-service.test.ts [hunk 4]
```diff
@@ -0,0 +1,698 @@
...
+  getCurrentBranch,
+  getMainBranch,
+  autoCommitCurrentBranch,
+  commit,
+  RUNTIME_EXCLUSION_PATHS,
+  VALID_BRANCH_NAME,
+  type GitPreferences,
+  type CommitOptions,
+  type TaskCommitContext,
+} from "../git-service.js";
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/kata/tests/git-service.test.ts:22]: Added a full `mergeSliceToMain` test suite with three cases:

- **Happy path**: creates a slice branch with one commit, calls `mergeSliceToMain`, asserts the squash commit appears on `main` with the correct conventional message, and verifies we land on `main`
- **Empty-slice guard**: creates a slice with no new commits, asserts the new `throw` path fires with `"staged nothing"` in the message
- **Conflict scenario**: creates conflicting edits on both `main` and the slice branch, asserts `MergeConflictError` is thrown with populated `conflictedFiles`, correct `strategy: "squash"`, and correct branch names

`MergeConflictError` is now imported and exercised. 78 tests passing.

File: apps/cli/src/resources/extensions/kata/git-service.ts [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+ */
+export const VALID_BRANCH_NAME = /^[a-zA-Z0-9._\-\/]+$/;
+
+/**
+ * Kata runtime paths that should be excluded from smart staging.
+ * These are transient/generated artifacts that should never be committed.
+ * All entries use the .kata-cli/ prefix (not .gsd/).
+ */
+export const RUNTIME_EXCLUSION_PATHS: readonly string[] = [
+  ".kata-cli/activity/",
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/kata/git-service.ts:96]: This is worth clarifying. The `RUNTIME_EXCLUSION_PATHS` intentionally use `.kata-cli/` prefixes — this was the change made during the port from gsd-pi (where paths were `.gsd/`). The rationale: the Kata CLI writes its own runtime artifacts (lock files, metrics, activity logs, worktrees) under `.kata-cli/` within the project directory, distinct from `.kata/` which is the planning artifact store (roadmaps, summaries, decisions).

However, you're right to flag it: the KATA-WORKFLOW.md documents that auto-mode saves session logs to `.kata/activity/`, not `.kata-cli/activity/`. If the activity log path diverges between documentation and the actual extension runtime, `.kata/activity/` files could leak into commits. That's a valid concern — but confirming the actual write paths requires checking the broader extension code outside this PR's scope. I'll track it as a follow-up to verify whether `.kata/` paths need to be added to (or substituted for) the exclusion list.

File: apps/cli/src/resources/extensions/kata/git-service.ts [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+      message: `Merge conflict while squashing ${sliceBranch} into ${mainBranch}`,
+      conflictedFiles,
+      strategy: prefs?.merge_strategy ?? "squash",
+      branch: sliceBranch,
+      mainBranch,
+    });
+  }
+
+  const squashMsg = `feat(${milestoneId}/${sliceId}): ${sliceTitle}`;
+  commit(basePath, { message: squashMsg });
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/kata/git-service.ts]: Fixed — same root cause as the Greptile thread on this line. `commit()` return value is now checked and we throw when `committed === null`. See that thread for details.

File: apps/cli/src/resources/extensions/kata/git-service.ts [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+  prefs?: GitPreferences,
+): MergeSliceResult {
+  const sliceBranch = getCurrentBranch(basePath);
+  const mainBranch = getMainBranch(basePath, prefs);
+
+  runGit(basePath, ["switch", mainBranch]);
+
+  try {
+    runGit(basePath, ["merge", "--squash", sliceBranch]);
+  } catch {
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/kata/git-service.ts]: Fixed. The catch block now checks `conflictedFiles.length === 0` before throwing `MergeConflictError`, and rethrows the original git error for non-conflict failures:

```ts
} catch (mergeErr) {
  const conflictedFiles = runGit(
    basePath,
    ["diff", "--name-only", "--diff-filter=U"],
    { allowFailure: true },
  )
    .split("\n")
    .filter(Boolean);

  // Non-conflict failure (missing branch, dirty index, invalid ref, etc.)
  // — rethrow the original git error rather than misclassifying it as a conflict.
  if (conflictedFiles.length === 0) {
    throw mergeErr;
  }

  throw new MergeConflictError({ ... });
}
```

`MergeConflictError` is now only thrown when there are actual unmerged paths.

File: apps/cli/src/resources/extensions/kata/git-service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/cli/src/resources/extensions/kata/git-service.ts:499]: Fixed — same issue as the Greptile and Codex threads on this block. Chose the throw approach over the empty-string return: an empty squash is an error condition in Kata's squash-merge workflow, not a valid no-op. A caller that gets a `MergeSliceResult` back should be able to trust a real commit was created.
